cmd/goal/box.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var boxName string
+
+func init() {
+	appCmd.AddCommand(appBoxCmd)
+
+	appBoxCmd.AddCommand(appBoxInfoCmd)
+	appBoxCmd.AddCommand(appBoxListCmd)
+	appBoxCmd.PersistentFlags().Uint64Var(&appIdx, "app-id", 0, "Application ID")
+	appBoxCmd.MarkFlagRequired("app-id")
+
+	appBoxInfoCmd.Flags().StringVarP(&boxName, "name", "n", "", "Application box name. Use the same form as app-arg to name the box.")
+	appBoxInfoCmd.MarkFlagRequired("name")
+}
+
+var appBoxCmd = &cobra.Command{
+	Use:   "box",
+	Short: "Read application box data",
+	Args:  cobra.ArbitraryArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		// If no arguments passed, we should fallback to help
+		cmd.HelpFunc()(cmd, args)
+	},
+}
+
+var appBoxInfoCmd = &cobra.Command{
+	Use:   "info",
+	Short: "Retrieve information about an application box.",
+	Long:  "Retrieve information about an application box. The returned box name and value are base64 encoded.",
+	Args:  validateNoPosArgsFn,
+	Run: func(cmd *cobra.Command, args []string) {
+		_, client := getDataDirAndClient()
+
+		// Ensure box name is specified
+		if boxName == "" {
+			reportErrorf(errorMissingBoxName)
+		}
+
+		// Get box info
+		box, err := client.GetApplicationBoxByName(appIdx, boxName)
+		if err != nil {
+			if strings.Contains(err.Error(), fmt.Sprintf("{\"message\":\"box not found\"}")) {
```

Comment:
Is there any utility method in `daemon/algod/api/server/v2` for checking whether an error matches a particular error message? If not, is it worth adding one?

Reply:
I doubt it, the Go preferred way to do things like this is to use `Error.Is` or `Eror.As`, which works on the identity of the error object or type of the error, respectively, not the String.

Is the err returned identifiable in either of those ways?

Reply:
did a quick search, found this line: https://github.com/algorand/go-algorand/blob/feature/avm-box/daemon/algod/api/server/v2/handlers.go#L207.

The corresponding implementation: https://github.com/algorand/go-algorand/blob/feature/avm-box/daemon/algod/api/server/v2/handlers.go#L1179, and you may want to match against badRequest error, from https://github.com/algorand/go-algorand/blob/feature/avm-box/daemon/algod/api/server/v2/utils.go#L44?

Reply:
I think the error might be identifiable in one of those ways but the methods from the `deamon/algod/api/server/v2` package are not exported so I wouldn't be able to construct something for comparison purposes. The actual error I'm trying to test for is created on this line: https://github.com/algorand/go-algorand/blob/feature/avm-box/daemon/algod/api/server/v2/handlers.go#L1199.

Reply:
Another line I remembered: if we can compare against this error line: https://github.com/algorand/go-algorand/blob/feature/avm-box/daemon/algod/api/server/v2/errors.go#L24. which is used in previous handler, then we can "kinda" compare error message?

Reply:
> the problem is that errors.Is in the first handler has access to the error messages whereas cmd/goal does not.

ugh, then I guess to use `errors.Is`, one has to construct the error obj with corresponding error msg, which is not so plausible here... guess the only approach is to string match in error message...?

Reply:
That's the conclusion I came to as well. Might be worth adding some mechanism for this in the future, but that's outside the scope of this PR.